src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs

```diff
@@ -15,78 +15,118 @@ public class MonitoredEndpointsModule : ApiModule
 
                 var instances = GetMonitoredEndpointInstances(endpointRegistry, endpointName, activityTracker);
-                var period = ExtractHistoryPeriod();
 
-                FillInInstanceData(instances, criticalTimeStore, period, (e, v) => e.CriticalTime = v, IntervalsAggregator.AggregateTimings);
-                FillInInstanceData(instances, processingTimeStore, period, (e, v) => e.ProcessingTime = v, IntervalsAggregator.AggregateTimings);
-                FillInInstanceData(instances, retriesStore, period, (e, v) => e.Retries = v, IntervalsAggregator.AggregateRetries);
-                FillInInstanceData(instances, processingTimeStore, period, (e, v) => e.Throughput = v, intervals => IntervalsAggregator.AggregateTotalMeasurementsPerSecond(intervals, period));
+                foreach (var metric in instanceMetric)
+                {
+                    var store = metricByInstance.Single(s => s.GetType() == metric.StoreType);
+                    var intervals = store.GetIntervals(period, DateTime.UtcNow).ToLookup(k => k.Id.InstanceId);
 
-                return Negotiate.WithModel(instances);
-            };
-        }
+                    foreach (var instance in instances)
+                    {
+                        var values = metric.Aggregate(intervals[instance.Id].ToList(), period);
 
-        HistoryPeriod ExtractHistoryPeriod()
-        {
-            return HistoryPeriod.FromMinutes(Request.Query["history"] == null || Request.Query["history"] == "undefined" ? DefaultHistory : (int)Request.Query["history"]);
-        }
+                        instance.Metrics.Add(metric.ReturnName, values);
+                    }
+                }
 
-        static void FillInEndpointData(MonitoredEndpoint[] endpoints,
-            VariableHistoryIntervalStore store,
-            HistoryPeriod period,
-            Action<MonitoredEndpoint, MonitoredEndpointValues> setter,
-            Func<List<IntervalsStore.EndpointInstanceIntervals>, MonitoredEndpointValues> aggregate)
-        {
-            var intervals = store.GetIntervals(period, DateTime.UtcNow).ToLookup(k => k.Id.EndpointName);
+                var messageTypes = GetMonitoredMessageTypes(metricByMessageType, period, endpointName);
 
-            foreach (var endpoint in endpoints)
-            {
-                var values = aggregate(intervals[endpoint.Name].ToList());
+                foreach (var metric in messageTypeMetric)
+                {
+                    var store = metricByMessageType.Single(s => s.GetType() == metric.StoreType);
+                    var intervals = store.GetIntervals(period, DateTime.UtcNow).ToLookup(k => k.Id);
 
-                setter(endpoint, values);
-            }
-        }
+                    foreach (var messageType in messageTypes)
+                    {
+                        var values = metric.Aggregate(intervals[new EndpointMessageType(endpointName, messageType.MessageType)].ToList(), period);
```

Comment:
It looks quite complex: double loop, some `ToList()`. Should we consider a benchmark for it?

Reply:
I'll work on performance test and after that we can re-evaluate.